src/mesh/replicated_mesh.C

```diff
@@ -1417,5 +1417,151 @@ dof_id_type ReplicatedMesh::n_active_elem () const
                                                  this->active_elements_end()));
 }
 
+std::map<Point, std::vector<std::vector<Point>>>
+ReplicatedMesh::get_boundary_points() const
+{
+  // find number of disconnected subdomains
+  std::vector<Point> interior_points;
+  std::vector<subdomain_id_type> subdomains(max_elem_id() + 1, Elem::invalid_subdomain_id);
+
+  subdomain_id_type subdomain_counter = 0;
+
+  // a stack for visiting elements
+  std::vector<Elem *> list;
+  list.reserve(n_elem());
+
+  dof_id_type visited = 0;
+  do
+  {
+    for (const auto & elem : active_element_ptr_range())
+      if (subdomains[elem->id()] == Elem::invalid_subdomain_id)
+      {
+        list.push_back(elem);
+        interior_points.push_back(elem->centroid());
+        subdomains[elem->id()] = subdomain_counter;
+        break;
+      }
+    while (list.size() > 0)
+    {
+      // pop up an element
+      Elem * elem = list.back(); list.pop_back(); ++visited;
+
+      for (auto s : elem->side_index_range())
+      {
+        Elem * neighbor = elem->neighbor_ptr(s);
+        if (neighbor != nullptr && subdomains[neighbor->id()] == Elem::invalid_subdomain_id)
+        {
+          list.push_back(neighbor);
+          subdomains[neighbor->id()] = subdomain_counter;
+        }
+      }
+    }
+    std::cout << std::endl;
+    subdomain_counter++;
+  }
+  while (visited != n_elem());
+  // now subdomain_counter stores the number of disconnected subdomains
```

Comment:
Technically we've found the number of components that aren't connected through element sides.  You could also have e.g. a mesh with two elements connected at only one corner, and this algorithm would claim that those are two separate subdomains, even though topologically a connection at one point is still a connection.  Is that okay?

Reply:
Yes, subdomains connected through d-1 elems only are considered connected. We can actually separate this part of the code into another function which can have options to choose d-1 to 0 connection.